ext/mbstring/libmbfl/mbfl/mbfl_encoding.h

```diff
@@ -129,10 +130,92 @@ struct mbfl_convert_vtbl {
 	void (*filter_copy)(struct _mbfl_convert_filter *src, struct _mbfl_convert_filter *dest);
 };
 
-/*
- * encoding
- */
-typedef struct _mbfl_encoding {
+typedef struct {
+	unsigned char *out;
+	unsigned char *limit;
+	uint32_t state;
+	uint32_t errors;
+	uint32_t replacement_char;
+	unsigned int error_mode;
+	zend_string *str;
+} mb_convert_buf;
+
+typedef size_t (*mb_to_wchar_fn)(unsigned char **in, size_t *in_len, uint32_t *out, size_t out_len, unsigned int *state);
+typedef void (*mb_from_wchar_fn)(uint32_t *in, size_t in_len, mb_convert_buf *out, bool end);
+
+static inline void mb_convert_buf_init(mb_convert_buf *buf, size_t initsize, uint32_t repl_char, unsigned int err_mode)
+{
+	buf->state = buf->errors = 0;
+	buf->str = emalloc(_ZSTR_STRUCT_SIZE(initsize));
+	buf->out = (unsigned char*)ZSTR_VAL(buf->str);
+	buf->limit = buf->out + initsize;
+	buf->replacement_char = repl_char;
+	buf->error_mode = err_mode;
+}
+
+#define MB_CONVERT_BUF_ENSURE(buf, out, limit, needed) \
+	if ((limit - out) < (needed)) { \
+		size_t oldsize = limit - (unsigned char*)ZSTR_VAL(buf->str); \
+		size_t newsize = oldsize + MAX(oldsize >> 1, needed); \
```

Comment:
This growth strategy looks a bit aggressive: It's effectively doing a `*3` on each step, while `*2` would be more typical.

Reply:
I think you might have that backwards. `oldsize + (oldsize >> 1)` is 50% growth (`*0.5`).

Reply:
I indeed did have that backwards :)